0023-keymanager-secret-sharing.md

```diff
@@ -0,0 +1,671 @@
+
+  /// Return switch data point (integer) for the calling node.
+  /// 
+  /// The point is evaluation of the shared secret (polynomial)
+  /// at the given y value (caller id).
+  /// 
+  /// Switch point:
+  ///     P = g(caller_id) = B(node_id,caller_id)
+  ///
+  /// WARNING: This method must be called over a secure channel as
+  /// the polynomial needs to be kept secret and generated only
+  /// for authorized nodes.
+  fn switch_point(&self, churp_id: u8, round: u64) -> Integer;
+
+  /// Return the key share for the given round.
+  ///
+  /// To construct a derived key, the caller is required to collect 
+  /// at least the threshold number of key shares and combine them
+  /// locally to construct the derived key.
+  ///
+  /// Key share:
+  ///     K = H(runtime_id, key_pair_id)^{r g(0)}
+  ///
+  /// WARNING: This method must be called over a secure channel as
+  /// the key share needs to be kept secret and generated only
+  /// for authorized nodes.
+  fn key_share(&self, churp_id: u8, round: u64, runtime_id: Namespace, 
+    key_pair_id: KeyPairId) -> Integer;
+}
+```
+
+Methods:
+
+- `churp/init`
+
+- `churp/verification_matrix`
+
+- `churp/bivariate_share`
+
+- `churp/dealing`
+
+- `churp/share_reduction`
+
+- `churp/proactive_randomization`
+
+- `churp/full_share_distribution`
+
+- `churp/switch_point`
+
+- `churp/key_share`
+
+### Key manager client
+
+```rust
+/// Key manager client for CHURP.
+pub trait Client {
+  /// Return the key for the given round.
+  ///
+  /// The key is combined from key shares obtained from the key manager nodes
+  /// in the current CHURP committee.
+  fn key(&self, churp_id: u8, round: u64, runtime_id: Namespace, key_pair_id: KeyPairId) -> Vec<u8>;
+}
+```
+
+### Positive (CHURP)
+
+- High security, as the master secret is shared among key manager nodes.
+- Supports proactive randomization (share refresh).
+- Dynamic committees.
+
+### Negative (CHURP)
+
+- Slower than KDC.
+
+## Key Derivation Center
+
+Key Derivation Center (KDC) is a secret sharing scheme based on
+the verifiable secret sharing scheme (VSS) where every node possesses
+only a share of the master secret. To derive a key from the master secret,
+one needs to obtain at least t+1 threshold number of key shares from distinct
+nodes and reassemble them locally.
+
+```rust
+/// Key manager application that implements key derivation center (KDC).
+pub trait KDC {
+  /// TODO: Define methods if we decide to implement KDC also.
+}
+```
+
+### Positive (KDC)
+
+- High security, as the master secret is shared among key manager nodes.
+- Supports proactive randomization (share refresh).
+- Faster than CHURP.
+
+### Negative (KDC)
+
+- The number of key manager nodes that share a master secret is fixed and
+  cannot be changed once shares are generated.
+- Support for replicating a share to a specific node is needed.
+- A trusted party needs to select a secret, generate secret shares and
```

Comment:
how is this a negative?

Reply:
You need a trusted party. This is negative. In CHURP, all nodes contribute entropy, so if at least t+2 nodes generate the secret, t can be corrupted.

Reply:
What if I created the secret or am otherwise about to reconstruct it? It doesn't matter if I don't trust myself with the shares because I already have the secret anyway!

For simple cases like cdNFTs, the artist is trusted and really just wants an API for dealing the shares and allowing a smart contract to say who gets them later.